Allow for lazy registration of app components

jshcrowthe opened the pull request:
Fixes #14 

Live demo available here:
http://jsbin.com/zozuqeh/edit?js,console

mbleigh commented:
Well that's a surprisingly small change. I think the concern previously was cross-product hooks, particularly with Auth. Have you tried this out with loading the Database first, then Auth later?

So 1) load DB, read some data that doesn't require auth 2) load auth lazily, load some data from the DB that does require auth.

jshcrowthe commented:
@mbleigh So the cross-product hooks that you mentioned are all taken care of. However I did update my JSBin to demo your use case and discovered that the stub methods that `FirebaseApp` creates for the auth methods, needs to forward those methods on to the auth module if it is lazily instantiated. Should hopefully have that working in just a bit.

AFAICT none of the other products have this issue though as they all call into `getToken` at the time of the call whereas database needs to register token listeners.

justweb1 commented:
@jshcrowthe I was trying to figure out a way to auto inititializeApp when hosted on Firebase. I was trying this through Polymerfire but the issue @mbleigh brought up was lazy loading the elements. Would it be appropriate to add a test to your PR to see if `'/__/firebase/init.js'` and if it is there to import it and run initializeApp automatically?

jshcrowthe commented:
The reason why I followed this approach is largely based on how Firebase Auth currently registers itself with Firebase App.

Currently we are stubbing some auth-specific code that is required by certain modules (i.e. `storage`, `database`) in the event that auth does not exist (see https://github.com/firebase/firebase-js-sdk/blob/master/src/app/firebase_app.ts#L433-L448). This implies a dependency of these components upon those auth methods. As our source is currently spread across module implementations (some rely on imports and our webpack compilation, others rely on the `firebase` and `firebase.INTERNAL` namespaces), we have some workarounds in place to help bridge this gap.

IMO once we have access to the Auth source in this repo, a refactoring of this code and the associated database/storage code, could happen that would extract this specific module into a standalone piece. This piece would stub the API and provide each component (in this case auth) the ability to control how this functionality is "upgraded." That said, with how we currently distribute to CDNs, the "stubbing" module would be included with `firebase-app.js` as this code is shared across webpack entry points.

I will add a comment to explain the need for this and a "todo" to remove it once able?

jshcrowthe commented on src/app/firebase_app.ts:
```diff
@@ -229,27 +233,20 @@ class FirebaseAppImpl implements FirebaseApp {
-  private services_: {[name: string]:
-                          {[instance: string]: FirebaseService}} = {};
-  public INTERNAL: FirebaseAppInternals;
+  private _registeredServices: {
```
Good catch I'll remove this. It was an artifact of an old implementation.

jshcrowthe commented on src/app/firebase_app.ts:
```diff
@@ -288,32 +285,42 @@ class FirebaseAppImpl implements FirebaseApp {
    * on demand).
+   * @internal
    */
-  private getService(name: string, instanceString?: string): FirebaseService
```
This function wasn't removed, but renamed. With the new registration methodology, we need to call this function outside of the class which is prevented by the `private` keyword.

It was replaced with the internal `_getService` function that we call both inside of the class and outside.

jshcrowthe commented on src/app/firebase_app.ts:
```diff
@@ -474,36 +479,30 @@ export function createFirebaseNamespace(): FirebaseNamespace {
     if (appHook) {
       appHooks[name] = appHook;
+      getApps().forEach(app => {
+        appHook('create', app);
```
The two functions that currently will trigger the `create` hook for services are `initializeApp` and `registerService`. Though they are similar they are slightly different and actual have no overlap in terms of when appHooks are fired.

- `initializeApp` attempts to run appHooks for every factory currently in existence on the initialized app.
- `registerService` runs only the new appHook (if applicable) on all existing instances of `FirebaseApp` it then also registers itself as one of the available appHooks so that future `initializeApp` calls will trigger the hook.

The `create` hook will be run once, and only once, per instance of Firebase App. Because one can't register multiple services of the same name, attempting to register multiple appHooks under the same name, and trigger a duplicate function call, should fail with an error long before the call is ever attempted.

jshcrowthe commented:
@bojeil-google I have addressed some of your concerns and added some tests to verify that functionality stays functioning (turns out there was an alternate way to do it that I brought into compliance with the API).

jshcrowthe commented:
@justweb1 I don't know if this PR is the right place to add that type of functionality. That said, I think if we were to do something explicit (i.e. something like allowing people to call `firebase.initializeApp` w/o any args, or a separate function altogether) for this that'd work. 

However I can also see this type of functionality being a nice thing for an abstraction layer like Polymerfire, angularfire, etc.

Thoughts?

justweb1 commented:
The thought was originally for it to be built into Polymerfire and I had even started work on it. After talking with @mbleigh, it seems to make more sense to run it directly in the sdk. One of the main issues I was having was due to the fact I needed lazy loading. Then, when it boils down to it placing a check and calling the init.js directly from the sdk is a much better option and will require way less code. This will also simplify the initialization and improve user experience.

jshcrowthe commented:
@justweb1 let's open a separate issue for this and continue discussion there.

sphippen commented on src/app/firebase_app.ts:
```diff
@@ -474,36 +491,30 @@ export function createFirebaseNamespace(): FirebaseNamespace {
+     * otherwise, return a proxied reference to the same service
+     */
+    factories[name] = allowMultipleInstances ? createService :
+      (...args) => createService.apply(this, args);
```
Seems like this ternary statement isn't really doing anything? Before, the if statement was always sending DEFAULT_ENTRY_NAME instead of the passed instanceString if `!allowMultipleInstances`, but since you handle the multiple instance name thing further down now (inside the function assigned to `FirebaseAppImpl.prototype[name]`), you should just be able to do

    factories[name] = createService;
and have it work, right?

jshcrowthe replied:
Great point, I think I had meant to do this, but never actually did it. Fixed!

bojeil-google commented on src/app/firebase_app.ts:
```diff
@@ -516,6 +526,12 @@ export function createFirebaseNamespace(): FirebaseNamespace {
+    FirebaseAppImpl.prototype[name] = function(...args) {
+      const serviceFxn = this._getService.bind(this, name);
+      return serviceFxn.apply(this, allowMultipleInstances ? args : []);
+    }
```
It's late, maybe I'm misreading:
So this will patch in the service factory: app.storage(instanceString);
Will this pass instanceString to createService(app, extendApp, instanceString) ?

This statement, as quoted from above, doesn't pass the instance string.
let service = this.firebase_.INTERNAL.factories[name](this, this.extendApp.bind(this));

If you update the test below where I commented to confirm this, then you are good to go.

jshcrowthe replied:
Great catch on this. I have refactored to ensure that we are passing through the `instanceIdentifier` I must have missed it in adding code back.

In addition I have added an additional assertion to two of the tests below to ensure that we don't break this going forward.

bojeil-google commented on tests/app/unit/firebase_app.test.ts:
```diff
@@ -166,6 +166,113 @@ describe("Firebase App Class", () => {
+    const service2 = (firebase.app() as any).multiInstance(serviceIdentifier);
+    assert.strictEqual(service2, (firebase.app() as any).multiInstance(serviceIdentifier));
+
+    // Ensure that the two services **are not equal**
```
This doesn't check that the two difference services are getting their identifier passed to the underlying service constructor.
For Storage, this is:
function factory(app: FirebaseApp, unused: any, opt_url?: string): Service {
  return new Service(app, new XhrIoPool(), opt_url);
}
opt_url needs to be passed through.

jshcrowthe replied:
Gotcha, thanks!

jshcrowthe commented:
@bojeil-google I have responded to your last two comments. Thanks sir!

We should be good to go. I'll get this merged.

bojeil-google commented:
Looks good. Thanks!
